### ISSUE TITLE
[use-case] Drag events

### ISSUE THREAD

**Issue by caub** [tc39/proposal-observable#113]
Without Observables, I have this kind of code for moving/dragging elements:

``` js
function move(e, cb) {
    const up = e => {
        document.removeEventListener('mouseup', up);
        document.removeEventListener('mousemove', cb);
    };
    cb(e); // trigger it now also
    document.addEventListener('mouseup', up);
    document.addEventListener('mousemove', cb);
};
// then in jsx:
<div onMouseDown={e => {
    const {cx, cy} = getCenter(this.refs.bar);
    move( 
        e,
        ({clientX,clientY})=>this.setState({foo: [clientX-cx, clientY-cy]}) 
    )
}}/>
```

If I undertand well, it's rewritten with Observable like this:

``` js
function move(e){
    return new Observable( ({next}) => {
        const up = e => {
            document.removeEventListener('mouseup', up);
            document.removeEventListener('mousemove', next);
        };
        next(e); // trigger it now also
        document.addEventListener('mouseup', up);
        document.addEventListener('mousemove', next);
    })
}

<div onMouseDown={e=>{
    const {cx, cy} = getCenter(this.refs.bar);
    move(e).subscribe( 
        ({clientX,clientY})=>this.setState({foo: [clientX-cx, clientY-cy]})
    );
    // move(e).subscribe({
    //  next({clientX,clientY}){ this.setState({foo: [clientX-cx, clientY-cy]}) }
    // })
}}/>
```

The advantage of Observable, there, is to make this pattern more standard I guess

**Comment by acutmore** [comment]
Absolutely correct. The advantage of Observables is to create a common pattern.

Returning an observable instead of passing in a callback is also key to making small testable correct components and composing them together. i.e avoid callback hell.

They can help avoid bugs. addEventListener requires the same string and function to be passed to removeEventListener. One can imagine a bug where the string passed to removeEventListener has a typo. Wrapping the addEventListener+removeEventListener pattern in a function which creates an observable eliminates this bug.

```javascript
function fromEvent(element, eventName) {
    return new Observable(observer => {
        const next = event => observer.next(event);
        element.addEventListener(eventName, next);
        return () => element.removeEventListener(eventName, next);
    });
}
```

If we also create some other helper functions (or import a library which contains these already):

```javascript
// Only pass the first value to the observer, then unsubscribe
function firstValue(observable) {
    return new Observable(observer =>
        observable.subscribe({
            start(subscription) {
                this.subscription = subscription;
            },
            next(v) {
                observer.next(v);
                observer.complete();
                this.subscription.unsubscribe();
            },
            error(e) {
                observer.error(e);
            },
            complete() {
                observer.complete();
            }
        })
    );
}

// Unsubscribe from the `source` when there is a value from the `end`
function takeUntil(end, source) {
    return new Observable(observer => {
        const sourceSubscription = source.subscribe(observer);
        const endSubscription = firstValue(end).subscribe(
            _ => sourceSubscription.unsubscribe(),
            err => observer.error(err)
        );
        return () => {
            sourceSubscription.unsubscribe();
            endSubscription.unsubscribe();
        };
    });
}
```

Now we can reuse the above in many places with clearer intention with less opportunity for error. The 'move' function can just become:

```javascript
function move(e){
    const moves = fromEvent(e, 'mousemove');
    const ups = fromEvent(e, 'mouseup');
    return takeUntil(ups, moves);
}
```

I hope this demonstrates the benefit of Observables

**TLDR:**
One observable is not much smaller/different to the code it replaces. Only when larger programs are broken down into many smaller functions which return observables does the code change more dramatically.
